internal/redact/redact_test.go

```diff
@@ -0,0 +1,193 @@
+package redact
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"golang.org/x/text/transform"
+)
+
+func TestEmptySecretError(t *testing.T) {
+	r, err := NewTransformer("")
+	assert.Nil(t, r)
+	assert.ErrorIs(t, err, ErrEmptySecret)
+}
+
+func TestNoopWithNoSecrets(t *testing.T) {
+	r, err := NewTransformer()
+	if !assert.NoError(t, err) {
+		return
+	}
+
+	str := "Hello world!"
+
+	res, n, err := transform.String(r, str)
+	assert.Equal(t, str, res)
+	assert.Equal(t, len(str), n)
+	assert.NoError(t, err)
+}
+
+func TestNoopWithNoMatch(t *testing.T) {
+	r, err := NewTransformer("bogus")
+	if !assert.NoError(t, err) {
+		return
+	}
+
+	str := "Hello world!"
+
+	res, n, err := transform.String(r, str)
+	assert.Equal(t, str, res)
+	assert.Equal(t, len(str), n)
+	assert.NoError(t, err)
+}
+
+func TestRedact(t *testing.T) {
+	cases := []struct {
+		name     string
+		secrets  []string
+		input    string
+		expected string
+	}{
+		{
+			name:     "one secret",
+			secrets:  []string{"secret"},
+			input:    "Hello secret world!",
+			expected: "Hello *** world!",
+		},
+		{
+			name:     "many secrets",
+			secrets:  []string{"one", "two", "three"},
+			input:    "Hello one - two - three world!",
+			expected: "Hello *** - *** - *** world!",
+		},
+		{
+			name:     "repeated secrets",
+			secrets:  []string{"one", "two", "three"},
+			input:    "Hello one two one three two three world!",
+			expected: "Hello *** *** *** *** *** *** world!",
+		},
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			r, err := NewTransformer(c.secrets...)
+			if !assert.NoError(t, err) {
+				return
+			}
+
+			res, n, err := transform.String(r, c.input)
+			assert.Equal(t, c.expected, res)
+			assert.Equal(t, len(c.input), n)
+			assert.NoError(t, err)
+		})
+	}
+}
+
+func TestOverlapSecretAcrossWrites(t *testing.T) {
+	cases := []struct {
+		name     string
+		secrets  []string
+		writes   []string
+		expected string
+	}{
+		{
+			name:     "simple",
+			secrets:  []string{"beepboop"},
+			writes:   []string{"hello beep", "boop world"},
+			expected: "hello *** world",
+		},
+		{
+			name:     "repeated chars",
+			secrets:  []string{"AAAAA"},
+			writes:   []string{"hello AAA", "AA world"},
+			expected: "hello *** world",
+		},
+		{
+			name: "multiple matching secrets",
+			secrets: []string{
+				"boop",
+				"bboop",
+				"beboop",
+				"beeboop",
+				"beepboop",
+			},
+			writes:   []string{"hello beep", "boop world"},
+			expected: "hello *** world",
+		},
+		{
+			name: "multiple matching secrets reverse",
+			secrets: []string{
+				"beepboop",
+				"beeboop",
+				"beboop",
+				"bboop",
+				"boop",
+			},
+			writes:   []string{"hello beep", "boop world"},
+			expected: "hello *** world",
+		},
```

Comment:
remark: these tests seemingly do nothing. `boop` will never match even with the overlap, same with `beeboop` and so on. Not sure what I was trying to test here 🤔